### ISSUE TITLE
Crash on functions accepting callbacks & positional arguments to them typed with `TypeVarTuple`

### ISSUE THREAD

**Issue by Enegg** [microsoft/pyright#7141]
**Describe the bug**
While type checking code below pyright hangs indefinitely.

**Code or Screenshots**
```python
from collections import abc
from typing_extensions import TypeVar, TypeVarTuple, Unpack

T = TypeVar("T")
PosArgsT = TypeVarTuple("PosArgsT")

def call(func: abc.Callable[[Unpack[PosArgsT]], T], *args: Unpack[PosArgsT]) -> T:
    return func(*args)

def callback(num: int, /) -> int:
    return num

call(call, call, callback, 123)
```
Code tested in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.348&pythonVersion=3.10&code=GYJw9gtgBAxmA28CmMAuBLMA7AzldEADmCKlAIYBGMAsAFCiRSoCeh6WA5gPpIAeqJLky58REmQAqbJADVyIADRRphOQskBXQsmUBVLIXIwA1vXqSoAXhUz5IABQAiSU4CU9AApgcAQRCcOJY2quogWjpIzt5%2BAUHu5nQAJkjAsOSIDsCaWDAAXBTUAHQAwhnwVMgA2lUGRqZVMf6BkgC6rcptygBUCoEFdcYmjT7NQa1uUAC0AHwqefRQS1AgSKiaIFhQ2bkOvXEedPQpaTDllEMOWJoQBRyoygD0k7P4WKgLdMsraxtb1xBEnxrOlMmdEMpwfBIechsoAIwAJgAzIdCCB7g4%2BG4gA) gets no response from the server:

![image](https://github.com/microsoft/pyright/assets/42005170/e1a3b584-8041-42f7-a66e-a217434d01df)
(introduce any kind of error like `foo: str = 123`, it gets no response)

Pylance indefinitely analyzes the file, causing VSCode to have high CPU usage and constantly allocating more memory, until it eventually crashes.

<details>
<summary>Output log</summary>

```
2024-01-27 20:48:34.828 [info] [Info  - 8:48:34 PM] (18856) Background analysis(3) started

# file opened around 20:56

2024-01-27 21:03:46.484 [info] 
<--- Last few GCs --->

[18856:0000591C00340000]   919106 ms: Mark-Compact 1990.0 (2065.2) -> 1987.0 (2070.2) MB, 3812.95 / 0.00 ms  (average mu = 0.359, current mu = 0.188) allocation failure; scavenge might not succeed
[18856:0000591C00340000]   924837 ms: Mark-Compact 1994.7 (2070.2) -> 1990.6 (2073.7) MB, 5643.50 / 0.00 ms  (average mu = 0.174, current mu = 0.015) allocation failure; scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory

2024-01-27 21:03:46.599 [info]  1: 00007FF717BD2786 node::SetTracingController+80086
 2: 00007FF717BD2A13 node::OnFatalError+595
 3: 00007FF71A8C19B3 v8::Function::NewInstance+835
 4: 00007FF71A8C1943 v8::Function::NewInstance+723
 5: 00007FF71A98A367 v8::CppHeap::CollectGarbageInYoungGenerationForTesting+74791
 6: 00007FF7194D1594 v8::MicrotasksScope::PerformCheckpoint+543732
 7: 00007FF7194CC8ED v8::MicrotasksScope::PerformCheckpoint+524109

2024-01-27 21:03:46.599 [info]  8: 00007FF7194BCCE7 v8::MicrotasksScope::PerformCheckpoint+459591
 9: 00007FF7196C7C73 v8::MicrotasksScope::PerformCheckpoint+2601683
10: 00007FF7199E4D3A v8_inspector::String16::String16+2535050
11: 00007FF71994FDD2 v8_inspector::String16::String16+1924898
12: 00007FF77A88FFF9 

2024-01-27 21:03:46.950 [info] [Error - 9:03:46 PM] Server process exited with code 134.
2024-01-27 21:03:46.959 [info] [Info  - 9:03:46 PM] Connection to server got closed. Server will restart.
2024-01-27 21:03:46.959 [info] true
2024-01-27 21:03:48.403 [info] [Info  - 9:03:48 PM] (18192) Pylance language server 2024.1.101 (pyright version 1.1.348, commit cfb1de0c) starting
```

</details>

**VS Code extension or command-line**
VS Code extension - Pylance `v2024.1.101` (prerelease)

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro the problem as described. I'll investigate further.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.364.
